pkg/compactor/block_upload.go

```diff
@@ -420,3 +491,180 @@ func (r bodyReader) ObjectSize() (int64, error) {
+	if meta == nil {
+		return blockUploadNotStarted, nil, nil, err
+	}
+
+	v, err := c.loadValidation(ctx, userBkt, blockID)
+	if err != nil {
+		return blockStateUnknown, meta, nil, err
+	}
+	if v == nil {
+		return blockUploadInProgress, meta, nil, err
+	}
+	if v.Error != "" {
+		return blockValidationFailed, meta, v, err
+	}
+	if time.Since(time.UnixMilli(v.LastUpdate)) < 5*time.Minute {
+		return blockValidationInProgress, meta, v, nil
+	}
+	return blockValidationStale, meta, v, nil
+}
+
+func (c *MultitenantCompactor) loadUploadingMeta(ctx context.Context, userBkt objstore.Bucket, blockID ulid.ULID) (*metadata.Meta, error) {
+	r, err := userBkt.Get(ctx, path.Join(blockID.String(), uploadingMetaFilename))
+	if err != nil {
+		if userBkt.IsObjNotFoundErr(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	defer func() { _ = r.Close() }()
+
+	v := &metadata.Meta{}
+	err = json.NewDecoder(r).Decode(v)
+	if err != nil {
+		return nil, err
+	}
+
+	return v, nil
+}
+
+func (c *MultitenantCompactor) loadValidation(ctx context.Context, userBkt objstore.Bucket, blockID ulid.ULID) (*validationFile, error) {
+	r, err := userBkt.Get(ctx, path.Join(blockID.String(), validationFilename))
+	if err != nil {
+		if userBkt.IsObjNotFoundErr(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	defer func() { _ = r.Close() }()
+
+	v := &validationFile{}
+	err = json.NewDecoder(r).Decode(v)
+	if err != nil {
+		return nil, err
+	}
+
+	return v, nil
+}
+
+func (c *MultitenantCompactor) uploadValidationWithError(ctx context.Context, logger log.Logger, blockID ulid.ULID,
+	userBkt objstore.Bucket, errorStr string) error {
+	val := validationFile{
+		LastUpdate: time.Now().UnixMilli(),
+		Error:      errorStr,
+	}
+	dst := path.Join(blockID.String(), validationFilename)
+	level.Debug(logger).Log("msg", fmt.Sprintf("uploading %s to bucket", validationFilename), "dst", dst)
+	buf := bytes.NewBuffer(nil)
+	if err := json.NewEncoder(buf).Encode(val); err != nil {
+		return errors.Wrap(err, "failed to encode block metadata")
+	}
+	if err := userBkt.Upload(ctx, dst, buf); err != nil {
+		return errors.Wrapf(err, "failed uploading %s to bucket", validationFilename)
+	}
+
+	return nil
+}
+
+func (c *MultitenantCompactor) uploadValidation(ctx context.Context, logger log.Logger, blockID ulid.ULID,
+	userBkt objstore.Bucket) error {
+	return c.uploadValidationWithError(ctx, logger, blockID, userBkt, "")
+}
+
+func (c *MultitenantCompactor) periodicValidationUpdater(ctx context.Context, logger log.Logger, blockID ulid.ULID,
+	userBkt objstore.Bucket, ch chan string, wg *sync.WaitGroup) {
+	defer wg.Done()
+	for {
+		select {
+		case <-ch:
+			return
+		case <-time.After(1 * time.Minute):
+			break
+		}
+	}
+	if err := c.uploadValidation(ctx, logger, blockID, userBkt); err != nil {
+		level.Warn(logger).Log("msg", "error during periodic update of validation file", "blockID", blockID)
+	}
```

Comment:
If upload of validation file fails, we should abort the validation. Why? Because we risk that another validation for the same block will be started, and two validations will run at the same time.

```go
func (c *MultitenantCompactor) periodicValidationUpdater(ctx context.Context, logger log.Logger, blockID ulid.ULID,
	userBkt objstore.Bucket, ch chan string, wg *sync.WaitGroup, cancelFn func()) {
	defer wg.Done()

	for {
		select {
		case <-ch:
			return
		case <-time.After(1 * time.Minute):
			if err := c.uploadValidation(ctx, logger, blockID, userBkt); err != nil {
				level.Warn(logger).Log("msg", "error during periodic update of validation file", "blockID", blockID)
				cancelFn()
				return
			}
		}
	}
}
```

Caller of this function can create new context with cancellation function and pass this function as parameter:

```go
	ctx, cancel := context.WithCancel(ctx)
	defer cancel()
	go c.periodicValidationUpdater(ctx, logger, blockID, userBkt, ch, &wg, cancel)
```

This way when validation updater fails, it cancels the context, and it tells caller to stop.